src/main/java/org/jitsi/videobridge/IceUdpTransportManager.java

```diff
@@ -474,27 +457,320 @@ private void iceStreamPairChange(PropertyChangeEvent ev)
+        candidatePE.setNetwork(0);
+        candidatePE.setPriority(candidate.getPriority());
+
+        // Advertise 'tcp' candidates for which SSL is enabled as 'ssltcp'
+        // (although internally their transport protocol remains "tcp")
+        Transport transport = candidate.getTransport();
+        if (transport == Transport.TCP && candidate.isSSL())
+        {
+            transport = Transport.SSLTCP;
+        }
+        candidatePE.setProtocol(transport.toString());
+
+        if (transport == Transport.TCP || transport == Transport.SSLTCP)
+        {
+            candidatePE.setTcpType(candidate.getTcpType());
+        }
+
+        candidatePE.setType(
+                CandidateType.valueOf(candidate.getType().toString()));
+
+        TransportAddress transportAddress = candidate.getTransportAddress();
+
+        candidatePE.setIP(transportAddress.getHostAddress());
+        candidatePE.setPort(transportAddress.getPort());
+
+        TransportAddress relatedAddress = candidate.getRelatedAddress();
+
+        if (relatedAddress != null)
+        {
+            candidatePE.setRelAddr(relatedAddress.getHostAddress());
+            candidatePE.setRelPort(relatedAddress.getPort());
+        }
+
+        pe.addChildExtension(candidatePE);
+    }
+
+    private String generateCandidateID(LocalCandidate candidate)
+    {
+        StringBuilder candidateID = new StringBuilder();
+
+        candidateID.append(conference.getID());
+        candidateID.append(Long.toHexString(hashCode()));
+
+        Agent iceAgent
+                = candidate.getParentComponent().getParentStream().getParentAgent();
+
+        candidateID.append(Long.toHexString(iceAgent.hashCode()));
+        candidateID.append(Long.toHexString(iceAgent.getGeneration()));
+        candidateID.append(Long.toHexString(candidate.hashCode()));
+
+        return candidateID.toString();
+    }
+
+    private void iceStateChange(PropertyChangeEvent ev)
+    {
+        IceProcessingState oldState = (IceProcessingState) ev.getOldValue();
+        IceProcessingState newState = (IceProcessingState) ev.getNewValue();
+
+        logger.info(getLoggingId() + "ICE state changed old_state=" +
+                oldState + " new_state=" + newState);
+
+        // We should be using newState.isEstablished() here, but we see
+        // transitions from RUNNING to COMPLETED, which should not happen and
+        // when they happen the connection is not successful. So we handle that
+        // case separately below.
+        if (IceProcessingState.COMPLETED.equals(newState))
+        {
+            iceConnected = true;
+            onIceConnected();
+        }
+        else if (IceProcessingState.FAILED.equals(newState)
+            || (IceProcessingState.RUNNING.equals(oldState)
+                    && IceProcessingState.TERMINATED.equals(newState)))
+        {
+            onIceFailed();
+        }
+    }
+
+    protected void onIceConnected()
```

Comment:
looks like there is logic in `IceDtlsTransportManager#onIceConnected` that could be here.  saving that for later?

Reply:
The iceConnectedProcessed check? I don't mind it in IceDtls (IceUdp is getting fat).